Quite a bit of stuff for updating and fixing P1684 (#216)
ORNL/cpp-proposals-pub: P1684-mdarray/P1684.md

#### Changes from R1

- update and reword non-wording text to harmonize with P0009R16
- fix synopsis missing Alloc argument in some places
- fix constraints on some constructors to allow std::array as container

> - add missing wording for std::initializer list

Reviewer: ```suggestion
- add missing wording for constructors from`std::initializer_list`
```

> - add missing wording for constructors from`std::initializer_list`

Reviewer: missing space
```suggestion
- add missing wording for constructors from `std::initializer_list`
```

> - add constructors from ranges

Reviewer: They asked for this? :/

Author: more or less, they asked what about those, and both container adaptors and containers have them so ...

> - always use std::vector as the default container type, based on LWEG guidance

Reviewer 2: ```suggestion
- always use `std::vector` as the default container type, based on LEWG guidance
```

- remove container accessor function

>   - its a bad idea to allow someone to resize the container owned by an mdarray ...

Reviewer 2: ```suggestion
  - it's a bad idea to allow someone to resize the container owned by an `mdarray` ...
```

Reviewer: Essentially giving the user a chance to break the class invariance

      const mdarray<OtherElementType, OtherExtents, 
                   OtherLayoutPolicy, OtherContainer>& other);

  // [mdarray.ctors.alloc], mdarray constructors with allocators

>   template<class Alloc>
>     constexpr mdarray(const Extents& ext, const Alloc& a);

Reviewer: Why do we have both a version with and without allocator?
Can't we do
```suggestion
  template<class Allocator = typename container_type::allocator_type>
    constexpr mdarray(const Extents& ext, const Allocator& alloc = Allocator());
```

Reviewer: Christian pointed out that it does not work if underlying container is `std::array`

  template<class Alloc>
    constexpr mdarray(const mapping_type& m, const Alloc& a);

  template<class Alloc>
    constexpr mdarray(const container_type& c, const extents_type& ext, const Alloc& a);
  template<class Alloc>
    constexpr mdarray(const container_type& c, const mapping_type& m, const Alloc& a);

  template<class Alloc>
    constexpr mdarray(container_type&& c, const extents_type& ext, const Alloc& a);
  template<class Alloc>
    constexpr mdarray(container_type&& c, const mapping_type& m, const Alloc& a);

  template<class Alloc>
    constexpr mdarray(initializer_list<element_type> init, const extents_type& ext, const Alloc& a);
  template<class Alloc>
    constexpr mdarray(initializer_list<element_type> init, const mapping_type& m, const Alloc& a);
  
  template<@_container-compatible-range_@<element_type> R, class Alloc>

>     constexpr mdarray(from_range_t, R&& rg, const extents_type& ext);

Reviewer: ```suggestion
    constexpr mdarray(from_range_t, R&& rg, const extents_type& ext, const Alloc& a);
```

  template<@_container-compatible-range_@<element_type> R, class Alloc>

>     constexpr mdarray(from_range_t, R&& rg, const mapping_type& m);

Reviewer: ```suggestion
    constexpr mdarray(from_range_t, R&& rg, const mapping_type&, const Alloc& a);
```

   * [1.3]{.pnum} `is_constructible_v<mapping_type, Extents>` is `true`, and

   * [1.4]{.pnum} If `container_type` is not a specialization of `array`, `is_constructible_v<container_type, size_t>` is `true`.

> [3]{.pnum} *Preconditions:* If `container_type` is a specialization of `array`, `mapping_type(Extents(static_cast<size_type>(std::move(exts))...)).required_span_size() >= size(container_type())` is `true`.

Reviewer: there is no [2]

constexpr mdarray(container_type&& c, const mapping_type& m);
```

[20]{.pnum} *Preconditions:* `c.size() >= m.required_span_size()` is `true`.

[21]{.pnum} *Effects:*

   * [21.1]{.pnum} Direct-non-list-initializes _`map_`_ with `m`, and

   * [22.2]{.pnum} Direct-non-list-initializes `ctr_` with `std::move(c)`.

```c++

> template<class... SizeTypes>

Reviewer 2: Do we need to consider the special case of `array` specializations here, as in the above constructors?

Author: no, this constructor requires constructibility from initializer list, i.e. its a forwarding constructor.

  explicit constexpr mdarray(std::initializer_list<element_type> init, SizeTypes... exts);
```

[7]{.pnum} *Constraints:*

   * [7.1]{.pnum} `(is_convertible_v<SizeTypes, size_type> && ...)` is `true`,

   * [7.2]{.pnum} `is_constructible_v<extents_type, static_cast<size_type>(SizeTypes)...>` is `true`,

   * [7.3]{.pnum} `is_constructible_v<mapping_type, extents_type>` is `true`, and

   * [7.4]{.pnum} `is_constructible_v<container_type, std::initializer_list<element_type>>` is `true`.

> [8]{.pnum} *Preconditions:* `init.size() >= mapping_type(extents_type(static_cast<size_type>(std::move(exts))...)).required_span_size()` is `true`.

Reviewer 2: What would it mean for `init.size()` to be greater than the required span size from `exts...`?  Would that just cut off the "extra" elements of the `initializer_list`?

```c++
template<class Alloc>
  constexpr mdarray(container_type&& c, const mapping_type& m, const Alloc& a);
```

[41]{.pnum} *Constraints:* `is_constructible_v<container_type, container_type, Alloc>` is `true`.

[42]{.pnum} *Preconditions:* `c.size() >= m.required_span_size()` is `true`.

[43]{.pnum} *Effects:*

   * [43.1]{.pnum} Direct-non-list-initializes _`map_`_ with `m`, and

>    * [43.2]{.pnum} Direct-non-list-initializes `ctr_` with `std::move(c)` as the first argumnet and `a` as the second argument.

Reviewer 2: ```suggestion
   * [43.2]{.pnum} Direct-non-list-initializes `ctr_` with `std::move(c)` as the first argument and `a` as the second argument.
```
